lib/ansible/executor/task_executor.py

```diff
@@ -526,6 +526,10 @@ def _execute(self, variables=None):
             display.debug("running the handler")
             try:
                 result = self._handler.run(task_vars=variables)
+            except AnsibleActionSkip as e:
+                return dict(skipped=True, msg=to_text(e))
+            except AnsibleActionFail as e:
+                return dict(failed=True, msg=to_text(e))
```

Comment:
Depends... is ConnectionFailure flow control or recoverable failure?  If people writing action plugins (or the methods backing them) are allowed to use it as an alternative to return with results dict then it should also be included here (and its parent class hanged as well).